lib/services/s3.js

```diff
@@ -306,6 +341,42 @@ AWS.util.update(AWS.S3.prototype, {
   },
 
   /**
+   * Updates httpRequest with region. If region is not provided, then
+   * the httpRequest will be updated based on httpRequest.region
+   *
+   * @api private
+   */
+  updateReqBucketRegion: function updateReqBucketRegion(request, region) {
+    var httpRequest = request.httpRequest;
+    if (typeof region === 'string' && region.length) {
+      httpRequest.region = region;
+    }
+    if (!httpRequest.endpoint.host.match(/s3(?!-accelerate).*\.amazonaws\.com$/)) {
+      return;
+    }
+    var service = request.service;
+    var s3Config = service.config;
+    var s3BucketEndpoint = s3Config.s3BucketEndpoint;
+    if (s3BucketEndpoint) {
+      delete s3Config.s3BucketEndpoint;
+    }
+    var newConfig = AWS.util.copy(s3Config);
+    delete newConfig.endpoint;
+    newConfig.region = httpRequest.region;
+
+    var endpoint = (new AWS.S3(newConfig)).endpoint;
+    httpRequest.endpoint = new AWS.Endpoint(endpoint);
```

Comment:
The original `httpRequest.endpoint` was constructed by passing the service object endpoint into the `AWS.Endpoint` constructor. This essentially copies the non-hidden properties of the service object endpoint to a new object, which won't contain any of the hidden properties of the original. If the only purpose of constructing a new endpoint is to create a new copy so it's independent from the service object endpoint, then it won't be necessary to do it here, but I'll have to test to make sure the hidden properties don't make a difference.